### ISSUE TITLE
Thread safety concerns due to "memory barrier" pattern in value classes

### ISSUE THREAD

**Issue by Calvin-L** [tlaplus/tlaplus#439]
Many of the `Value` classes contain instances of a pattern that I think is an incorrect attempt at creating a "memory barrier". This pattern is [known to be incorrect in Java](https://www.cs.umd.edu/~pugh/java/memoryModel/BidirectionalMemoryBarrier.html).

The essence of the incorrect memory barrier pattern is
 - Writers acquire a lock, initialize an object, release the lock, re-acquire the lock, then write the initialized object to a place where readers can see it. Finally, they release the lock again.
 - Readers do not use any synchronization.

This pattern is seductive because it _does_ constrain the order in which writes reach main memory. There is also no issue if all reads and writes happen in the same thread. However, it is incorrect when there are concurrent writers and readers because the readers do not have to observe those writes in order. Readers can observe writes out-of-order if they page-in the reference to the initialized object before they page-in the writes that initialize the fields of the object.

Here is an example from `FcnRcdValue`. The "writer" pathway is in `createIndex()`:

```java
int[] tbl = new int[len];
Arrays.fill(tbl, -1);

synchronized(this) {
  for (int i = 0; i < this.domain.length; i++) {
    int loc = (this.domain[i].hashCode() & 0x7FFFFFFF) % len;
    while (tbl[loc] != -1) {
      loc = (loc + 1) % len;
    }
    tbl[loc] = i;
  }
}

synchronized(this) { this.indexTbl = tbl; }
```

The first `synchronized` block initializes the cells of the local array `tbl`, and the second one writes `tbl` to `this.indexTbl` where it can be picked up by readers.

The reader pathway is in `select(arg)`:

```java
if (this.indexTbl != null) {
  return this.values[this.lookupIndex(arg)];
}
if (this.isNorm) this.createIndex();

if (this.indexTbl != null) {
  return this.values[this.lookupIndex(arg)];
}
```

In this example, the reader's first check (`this.indexTbl != null`) can be true even when the cells of the `indexTbl` array are uninitialized (i.e. all zero). Even if the cells are initialized in main memory, those writes may not be visible to the reader yet. This could lead to readers crashing or returning incorrect values for function applications.

These classes contain instances of this pattern:
 - `FcnRcdValue`
 - `SetCapValue`
 - `SetCupValue`
 - `SetDiffValue`
 - `SetOfFcnsValue`
 - `SetOfRcdsValue`
 - `SetOfTuplesValue`
 - `SubsetValue`
 - `UnionValue`

Either this is a very subtle concurrency bug, or the synchronization is actually unnecessary and it can be removed to improve TLC's performance by one iota.

If this is a concurrency bug, I believe it can be fixed with careful use of the `volatile` keyword and removal (or merging) of some `synchronized` blocks.

I have never seen problems due to this code. However, I believe it is worth reporting and fixing.

**Comment by craft095** [comment]
This seems to be dangerous in context of modeling of safety critical applications, where we are using TLA+ and TLC:
- it may result in returning "success" even if some properties are violated
- it is difficult to argue that probability of such event is really low
- it has no obvious workaround

I believe it is important to fix the issue. Fix seems to be straightforward and I could do it, if you think it is reasonable.

**Comment by Calvin-L** [comment]
I'm certain nobody will object to the help. :)

Before you begin, you should know that the `Value` classes have an undocumented lifecycle that greatly reduces the need for synchronization. See e.g. the lengthy discussion in #361.

**Comment by lemmy** [comment]
Thanks, @craft095, for stepping up. :-)

 As @Calvin-L writes, given #361, #439 is assumed to be a non-issue. #439 was left open to be confirmed as such.

**Comment by craft095** [comment]
Awesome! @Calvin-L , @lemmy thank you for fixing it! BTW, the fix was committed two years ago, do you still think additional confirmation is required or it can be closed? Could I help with it somehow?

**Comment by lemmy** [comment]
@craft095 If you have the time, it would be greatly appreciated if you could re-confirm @Calvin-L and my reasoning here and in #361, so we can finally close #439. :-)

**Comment by Calvin-L** [comment]
For what it's worth, here is my assessment:

 - The `FcnRcdValue` code described above is _not_ harmless. I really do think there is a synchronization problem.
 - I am not sure if it is possible to demonstrate the issue on any actual JVM. Even so, I think we should implement a fix; TLC should not rely on implementation details of the JVM to protect against threading mishaps.
 - I could probably write a TLA+ spec to demonstrate the problem.
 - I think calling `createIndex()` after `normalize()` in `FcnRcdValue.fingerPrint()` should be sufficient to resolve the problem. The `synchronized` blocks can then be removed.
 - The `synchronized` blocks in the set classes (`Set*`, `SubsetValue`, `UnionValue`) look harmless and can be removed. They all appear in `convertAndCache()`, which is performed during fingerprinting, so they will not have any effect once those values are shared with other threads.

**Comment by Calvin-L** [comment]
TLA+ spec demonstrating the issue:
https://gist.github.com/Calvin-L/e2e9de865a9c89640c4a381738963278

**Comment by lemmy** [comment]
@Calvin-L I won't have a chance to look at your spec before next week.  If #361 does not take care of #439, we could also consider removing `createIndex` after checking if real-world specs rely on it.  This is motivated by me remembering running into a (performance) bottleneck caused by `createIndex` when simulating a spec with larger sets of values common during exhaustive checking.

**Comment by Calvin-L** [comment]
Removing `createIndex()` should also resolve the problem, and if it makes TLC a little more scalable then even better. :)

**Comment by lemmy** [comment]
For a few years, we ran the specs/models in https://github.com/tlaplus/tlaplus/tree/master/general/performance as performance and scalability tests. Unfortunately, the underlying infrastructure is now defunct.  For this issue, it would be good to get numbers with and without `createIndex`.  @craft095 Do you have the time to collect these numbers?

**Comment by craft095** [comment]
Sure I will do.

**Comment by craft095** [comment]
To do measurements, I have fixed [measure.sh and Grid5k](https://gist.github.com/craft095/7eef37c382a4f06df632857e970e088b)
Without these fixes createIndex is not triggrered.

Ubuntu 20.04, 32G RAM, i9-10900X CPU @ 3.70GHz, 10 physical cores

Summary (# of workers, AvgTimeDiff %):
* 20: -1.0%
* 10: 1.7%
* 5: 5.2%
* 2: 3.1%
* 1: 3.0%

AvgTimeDiff = (AvgWithCreateIndex - AvgWithoutCreateIndex)/AvgWithCreateIndex

More details:

With createIndex():
workers,run1,run2,run3,avg(sec)
20,97.28,94.56,95.61,95.8166666667
10,104.32,105.18,104.48,104.66
5,170.51,169.82,166.63,168.986666667
2,385.1,388.19,383.84,385.71
1,739.05,747.35,753.12,746.506666667

Without createIndex():
workers,run1,run2,run3,avg(sec)
20,100.56,94.5,95.37,96.81
10,102.87,103.68,102.22,102.923333333
5,158.51,160.41,161.85,160.256666667
2,368.59,378.24,374.21,373.68
1,722.58,725.31,725.14,724.343333333

**Comment by lemmy** [comment]
* None of the (real-world) specs in `general/performance/` or the https://examples.tlapl.us ever hit `createIndex()` with `domain.length > 10`
* The cost of `FcnRcdValue#equals` and `FcnRcdValue#hashCode` for `Grid5k` is not representative of real-world specs
* Index creation is expensive (`Value#hashCode` is implemented via `Value#fingerprint`) for short-lived FcnRcdValue instances generated as part of the evaluation of the next-state relation.  However, a constant FcnRcdValue that appears in state-, action-constraints, or invariants would likely benefit from an index
* `FcnRcdValue#select` could exploit the fact that `domain` will be sorted (`normalized`) => binary search

* Macro-benchmarks for a fixed number of workers (`16`) and `K=2` but varying values for `N` *and* an invariant that quantifies over the values of `Dom`. (Ubuntu 20.04, 32G RAM, i7-10875H CPU @ 2.30GHz):

With createIndex():
|N|run1|run2|run3|avg(sec)|
| --- | --|--|--|----- |
|15||||||
|25||||||

Without createIndex():
|workers|run1|run2|run3|avg(sec)|
| --- | --|--|--|----- |
|15||||||
|25|1103s|1111s|2640s|||

* Preliminary result: Microbenchmarks with JMH show no advantage of having an index table for functions up to 8192 elements.

**Comment by lemmy** [comment]
Thanks, @craft095 and @Calvin-L for the reviews!

**Comment by craft095** [comment]
Would it be possible to backport the fix to release version (1.7 at the moment)?

**Comment by lemmy** [comment]
There is the 1.7.1 branch from which we could create 1.7.2.  Would you want to go over the commits in `1.7.1..HEAD` and look for other candidate commits?  For example, there is an [incompatibility with the CommunityModules](https://github.com/tlaplus/CommunityModules/issues/54) that should be addressed.

On the other hand, we might also release 1.8.0.  Is this about certifying TLC as part of safety assurances?

**Comment by craft095** [comment]
That's right, it is about certifing TLC. Releasing 1.8.0 is definitely better option, if it does not take much time. What are your feeling whether it is possible to release 1.8.0 within a month?

**Comment by lemmy** [comment]
Does the certification extend to the Toolbox or is it limited to TLC?

**Comment by craft095** [comment]
Only TLC is considered.

**Comment by lemmy** [comment]
The closed and (trimmed) open issues in https://github.com/tlaplus/tlaplus/milestone/10 show that releasing 1.8.0 by the end of 2021 is not possible.  On the other hand, the [number of the closed issues since 1.7.1](https://github.com/tlaplus/tlaplus/milestone/10?closed=1) also make a 1.7.2 maintenance release a non-trivial effort.  @craft095 Is the end of 2021 a hard deadline for you?

**Comment by craft095** [comment]
No, it is not a hard deadline. The hard deadline may be late January/early February.
If 1.8.0 is not released by this time, would it be an acceptable solution to cherry-pick issues for critical fixes (#439, #362 - see below, why they are critical) to create 1.7.2?

Context:

Critical bugs are those which have following properties:
1. It is possible that TLC reports no violation, while there is violation
2. It is hard to propose recommendations/guidelines to work around the issue

I.e. critical bugs may introduce errors, which are left undetected.

I thought there was only #439, but after some tests I realized that #362 is also critical.
